### ISSUE TITLE
Native Apache Pulsar Support

### ISSUE THREAD

**Issue by sido420** [MaterializeInc/materialize#3517]
I was wondering if there is native support available for Apache Pulsar. 

If not, is there a tutorial on how to use Materialize to process MySQL --> Debezium --> Pulsar (Kafka layer) source? I know that Materialize works with Kafka streams and there is a Kafka connector for Pulsar but according to https://news.ycombinator.com/item?id=22364353, it's only available for Java clients.

In any case, it's evident that other folks are also looking for similar feature/support.

I am working on a proof of concept project using Materialize and would appreciate any guides, tutorials etc.

**Comment by sido420** [comment]
Can someone tell me the Kafka client version being used with Materialized? 

I tried the billing demo, which perhaps creates a Kafka source in Materialized. And, now it looks like every time Materialized starts it queries Kafka brokers which throws errors (my setup's issue thats why I need Kafka client version).

**Comment by benesch** [comment]
> Can someone tell me the Kafka client version being used with Materialized?

Run `materialized -vv` to see the version of librdkafka linked into your binary:

```
$ materialized -vv
materialized v0.3.2-dev (245172d228133c5ec582b75df66c440d0ad55837)
OpenSSL 1.1.1g  21 Apr 2020
librdkafka v1.4.2
```

Note that materialized uses librdkafka, not the Java client library.

I'll let @awang speak to our plans for Pulsar support.

**Comment by sido420** [comment]
@awang @benesch if you guys can figure out how to use Materialize with Kafka on Pulsar then we may have a quick solution. I opened an issue there https://github.com/streamnative/kop/issues/166

**Comment by sijie** [comment]
I am wondering what are the requirements for Materialize to support Pulsar natively.

**Comment by rjnn** [comment]
I believe the requirements would be for Pulsar to work with `librdkafka v1.4.2`, but @benesch is the authority here.

**Comment by sijie** [comment]
@rjnn oh, I was thinking of native support of Pulsar not through the Kafka-on-Pulsar project. Pulsar has its c++ library. Is that sufficient for making native support of Pulsar?

**Comment by frankmcsherry** [comment]
> I am wondering what are the requirements for Materialize to support Pulsar natively.

I think realistically there is some clean-up work for us before supporting other sources (or any other not-drop-in-replacement fabric), followed by someone with Pulsar experience writing some unknown amount of code. I'm not aware of technical limitations, just hours-in-the-day limitations.

Materialize doesn't have particularly complicated requirements for getting data in or out (read about them [here](https://github.com/MaterializeInc/materialize/blob/master/doc/developer/change_data_capture.md)). But (and I'm saying this without any Pulsar experience) each of the fabrics out there have quirks that make it non-trivial to just drop in support. Ideally we would get to a point where it would be easier for Pulsar folks to add support than for us to add support.

**Comment by sido420** [comment]
Kafka handler for Pulsar supports librdkafka v1.2.2.

Would **current** version of Materialize build fine with version v1.2.2 of `librdkafka`? 

- If yes, how do I build Materialize with v1.2.2?
- If no, which version of Materialize works with v1.2.2? I will start my Materialize journey with an older version in that case.
- If unsure, how do we build Materialize with v1.2.2? 

I built current version using the instructions in documentation but that came with v1.4.2.

**Comment by benesch** [comment]
You're welcome to try to build Materialize with v1.2.2, but we're not going to be able to support you in that endeavor, I'm afraid. It would be best if you could get Kafka-on-Pulsar to support the latest version of the Kafka API so that it is compatible with newer versions of librdkafka. We'll be upgrading to 1.5 shortly, too.

I don't think we presently have the engineering resources to invest in making these changes to KoP ourselves, but if there is enough interest, we could consider looking into it.

**Comment by sido420** [comment]
>> You're welcome to try to build Materialize with v1.2.2

I have Materialize source already. I will download v1.2.2 from https://github.com/edenhill/librdkafka/releases/tag/v1.2.2. How do I link Materialize with librdkafka v1.2.2 and build `materialized` binary?

>> but we're not going to be able to support you in that endeavor, I'm afraid
Thats fine. I am still on a pre-POC stage. 

Thanks.

**Comment by sido420** [comment]
Did some more research and found out that Materialize is using rust-rdkafka which is based on librdkafka.

I searched the rust-rdkafka repo and found out that this version https://github.com/fede1024/rust-rdkafka/tree/v0.22.0 is based on v1.2.2 of librdkafka.

**Comment by benesch** [comment]
Indeed, but there have been a number of breaking changes to the rust-rdkafka project in the meantime, and it won't be trivial to adjust the current version of Materialize to build against v0.22 of rust-rdkafka. (I'm working to separate the versioning of rust-rdkafka and rdkafka-sys, so that swapping out versions of librdkafka is easier, but we're not there yet.)

I really think it'll be easier to attack this problem from the KoP angle. Note that librdkafka supports connecting to old Kafka brokers, all the way back to pre-1.0 Kafka brokers. KoP doesn't actually need to support the newest APIs; it just needs to not panic upon seeing a new version of an API that it doesn't support, but just return "unsupported" and masquerade as an older version of Kafka.

**Comment by aahmed-se** [comment]
there is also a native rust pulsar client
https://github.com/wyyerd/pulsar-rs

**Comment by sploiselle** [comment]
No, we just have not received much signal for building out support for this and are retiring the issue. If we receive more signal that this is something many folks are interested in, we'll create a new issue.

**Comment by benesch** [comment]
I'm actually going to reopen this for tracking purposes.
